subcommands/keys/tuf_updates_add_offline_key.go

```diff
@@ -0,0 +1,124 @@
+	tufUpdatesCmd.AddCommand(add)
+}
+
+func doTufUpdatesAddOfflineKey(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	txid, _ := cmd.Flags().GetString("txid")
+	keyTypeStr, _ := cmd.Flags().GetString("key-type")
+	keyType := ParseTufKeyType(keyTypeStr)
+	keysFile, _ := cmd.Flags().GetString("keys")
+	roleName, _ := cmd.Flags().GetString("role")
+	roleName = ParseTufRoleNameOffline(roleName)
+
+	var creds OfflineCreds
+	if _, err := os.Stat(keysFile); err == nil {
+		creds, err = GetOfflineCreds(keysFile)
+		subcommands.DieNotNil(err)
+		subcommands.AssertWritable(keysFile)
+	} else if errors.Is(err, fs.ErrNotExist) {
+		creds = make(OfflineCreds, 0)
+		saveTufCreds(keysFile, creds)
+	} else {
+		subcommands.DieNotNil(err)
+	}
+
+	updates, err := api.TufRootUpdatesGet(factory)
+	subcommands.DieNotNil(err)
+
+	_, newCiRoot := checkTufRootUpdatesStatus(updates, true)
+
+	switch roleName {
+	case tufRoleNameRoot:
+		addOfflineRootKey(newCiRoot, creds, keyType)
+	case tufRoleNameTargets:
+		onlineTargetsId := updates.Updated.OnlineKeys["targets"]
+		if onlineTargetsId == "" {
+			subcommands.DieNotNil(errors.New("Unable to find online target key for factory"))
+		}
+		addOfflineTargetsKey(newCiRoot, creds, keyType, onlineTargetsId)
+	default:
+		panic(fmt.Errorf("Unexpected role name: %s", roleName))
+	}
+
+	newCiRoot.Signatures = make([]tuf.Signature, 0)
+	removeUnusedTufKeys(newCiRoot)
+	newProdRoot := genProdTufRoot(newCiRoot)
+
+	fmt.Println("= Uploading new TUF root")
+	tmpFile := saveTempTufCreds(keysFile, creds)
+	err = api.TufRootUpdatesPut(factory, txid, newCiRoot, newProdRoot, nil)
+	handleTufRootUpdatesUpload(tmpFile, keysFile, err)
+}
+
+func addOfflineRootKey(root *client.AtsTufRoot, creds OfflineCreds, keyType TufKeyType) {
+	subcommands.DieNotNil(checkNoTufSigner(root, creds, root.Signed.Roles["root"].KeyIDs))
+
+	kp := genTufKeyPair(keyType)
+	root.Signed.Keys[kp.signer.Id] = kp.atsPub
+	root.Signed.Expires = time.Now().AddDate(1, 0, 0).UTC().Round(time.Second) // 1 year validity
+	root.Signed.Roles["root"].KeyIDs = append(root.Signed.Roles["root"].KeyIDs, kp.signer.Id)
+
+	base := "tufrepo/keys/fioctl-root-" + kp.signer.Id
+	creds[base+".pub"] = kp.atsPubBytes
+	creds[base+".sec"] = kp.atsPrivBytes
+	fmt.Println("= New root keyid:", kp.signer.Id)
+}
+
+func addOfflineTargetsKey(root *client.AtsTufRoot, creds OfflineCreds, keyType TufKeyType, onlineTargetsId string) {
+	subcommands.DieNotNil(checkNoTufSigner(root, creds,
+		subcommands.SliceRemove(root.Signed.Roles["targets"].KeyIDs, onlineTargetsId)))
+
+	kp := genTufKeyPair(keyType)
+	root.Signed.Keys[kp.signer.Id] = kp.atsPub
+	root.Signed.Roles["targets"].KeyIDs = append(root.Signed.Roles["targets"].KeyIDs, kp.signer.Id)
+
+	base := "tufrepo/keys/fioctl-targets-" + kp.signer.Id
+	creds[base+".pub"] = kp.atsPubBytes
+	creds[base+".sec"] = kp.atsPrivBytes
+	fmt.Println("= New targets keyid:", kp.signer.Id)
```

Comment:
these two functions look a lot alike. I'm not against duplication if its easier or seemed cleaner to you. Just double checking.

Reply:
I copied a pattern from `rotateXXX` functions here.

At one point I was thinking about unification; but there are some tiny nuances that could become too vague during that process.  That's why I decided to do this _very explicitly_ and _maybe revisit later_.